openedx/core/djangoapps/waffle_utils/__init__.py

```diff
@@ -1,138 +1,44 @@
 """
-Utilities for waffle.
+Extra utilities for waffle: most classes are defined in edx_toggles.toggles, but we keep here some extra classes for
+usage within edx-platform. These classes cover course override use cases.
 
-Includes namespacing, caching, and course overrides for waffle flags.
 
 Usage:
 
-For Waffle Flags, first set up the namespace, and then create flags using the
-namespace.  For example::
-
    WAFFLE_FLAG_NAMESPACE = WaffleFlagNamespace(name='my_namespace')
-
    # Use CourseWaffleFlag when you are in the context of a course.
    SOME_COURSE_FLAG = CourseWaffleFlag(WAFFLE_FLAG_NAMESPACE, 'some_course_feature', __name__)
-   # Use WaffleFlag when outside the context of a course.
-   SOME_FLAG = WaffleFlag(WAFFLE_FLAG_NAMESPACE, 'some_feature', __name__)
 
-You can check these flags in code using the following::
+You can check this flag in code using the following::
 
-    SOME_FLAG.is_enabled()
     SOME_COURSE_FLAG.is_enabled(course_key)
 
-To test these WaffleFlags, see testutils.py.
-
-In the above examples, you will use Django Admin "waffle" section to configure
-for a flag named: my_namespace.some_course_feature
-
-You could also use the Django Admin "waffle_utils" section to configure a course
-override for this same flag (e.g. my_namespace.some_course_feature).
-
-For Waffle Switches, first set up the namespace, and then create the flag name.
-For example::
-
-    WAFFLE_SWITCHES = WaffleSwitchNamespace(name=WAFFLE_NAMESPACE)
-
-    ESTIMATE_FIRST_ATTEMPTED = 'estimate_first_attempted'
-
-You can then use the switch as follows::
-
-    WAFFLE_SWITCHES.is_enabled(waffle.ESTIMATE_FIRST_ATTEMPTED)
-
 To test WaffleSwitchNamespace, use the provided context managers.  For example:
 
     with WAFFLE_SWITCHES.override(waffle.ESTIMATE_FIRST_ATTEMPTED, active=True):
         ...
 
-For long-lived flags, you may want to change the default for devstack, sandboxes,
-or new Open edX releases. For help with this, see:
-openedx/core/djangoapps/waffle_utils/docs/decisions/0001-refactor-waffle-flag-default.rst
-
 Also see ``WAFFLE_FLAG_CUSTOM_ATTRIBUTES`` and docstring for _set_waffle_flag_attribute
 for temporarily instrumenting/monitoring waffle flag usage.
 
 """
-
 import logging
-import re
-import traceback
-from abc import ABCMeta
 from contextlib import contextmanager
-from weakref import WeakSet
 
-import crum
-import six
-from django.conf import settings
-from edx_django_utils.monitoring import set_custom_attribute
 from opaque_keys.edx.keys import CourseKey
-from waffle import flag_is_active, switch_is_active
 
-from openedx.core.lib.cache_utils import get_cache as get_request_cache
+from edx_toggles.toggles import WaffleFlag as BaseWaffleFlag
+from edx_toggles.toggles import WaffleFlagNamespace
+from edx_toggles.toggles import WaffleSwitch as BaseWaffleSwitch
+from edx_toggles.toggles import WaffleSwitchNamespace as BaseWaffleSwitchNamespace
```

Comment:
> Did isort not complain?

I guess not... :thinking:

Reply:
When I move the `opaque_keys` import and run `isort openedx/core/djangoapps/waffle_utils/__init__.py` it moves it back into position...

Reply:
Could it be related to how the dependency is temporarily being brought in? Ignore for now, and hopefully this will complain later. I say hopefully, because if not, there would be something weird with edx-toggles going on.

Reply:
After some testing, it appears that this occurs just with isort==4.3.21, which is the version required by edx-platform.

```
➭ cat testisort.py 
import logging
from contextlib import contextmanager

from opaque_keys.edx.keys import CourseKey

from edx_toggles.toggles import WaffleFlag as BaseWaffleFlag
from edx_toggles.toggles import WaffleFlagNamespace
from edx_toggles.toggles import WaffleSwitch as BaseWaffleSwitch
from edx_toggles.toggles import WaffleSwitchNamespace as BaseWaffleSwitchNamespace
➭ isort --version           

/#######################################################################\

     `sMMy`
     .yyyy-                                                      `
    ##soos##                                                    ./o.
          `     ``..-..`         ``...`.``         `   ```` ``-ssso```
     .s:-y-   .+osssssso/.     ./ossss+:so+:`    :+o-`/osso:+sssssssso/
     .s::y-   osss+.``.``     -ssss+-.`-ossso`   ssssso/::..::+ssss:::.
     .s::y-   /ssss+//:-.`   `ssss+     `ssss+   sssso`       :ssss`
     .s::y-   `-/+oossssso/  `ssss/      sssso   ssss/        :ssss`
     .y-/y-       ````:ssss`  ossso.    :ssss:   ssss/        :ssss.
     `/so:`    `-//::/osss+   `+ssss+-/ossso:    /sso-        `osssso/.
       \/      `-/oooo++/-      .:/++:/++/-`      ..           `://++/.


         isort your Python imports for you so you don't have to

                            VERSION 4.3.21

\########################################################################/

➭ isort --check --diff testisort.py
➭ ~/.local/bin/isort --version

                 _                 _
                (_) ___  ___  _ __| |_
                | |/ _/ / _ \/ '__  _/
                | |\__ \/\_\/| |  | |_
                |_|\___/\___/\_/   \_/

      isort your imports, so you don't have to.

                    VERSION 5.6.4

➭ ~/.local/bin/isort --check --diff testisort.py 
ERROR: /home/data/regis/tmp/testisort.py Imports are incorrectly sorted and/or formatted.
--- /home/data/regis/tmp/testisort.py:before	2020-10-16 14:39:37.099681
+++ /home/data/regis/tmp/testisort.py:after	2020-10-16 14:42:29.078119
@@ -1,9 +1,9 @@
 import logging
 from contextlib import contextmanager
-
-from opaque_keys.edx.keys import CourseKey
 
 from edx_toggles.toggles import WaffleFlag as BaseWaffleFlag
 from edx_toggles.toggles import WaffleFlagNamespace
 from edx_toggles.toggles import WaffleSwitch as BaseWaffleSwitch
-from edx_toggles.toggles import WaffleSwitchNamespace as BaseWaffleSwitchNamespace+from edx_toggles.toggles import \
+    WaffleSwitchNamespace as BaseWaffleSwitchNamespace
```

Reply:
Odd. Not worth spending more time on. Thanks for checking.